Disable deployment target specific optimizations for inlinable functions

eeckstein opened the pull request:
The specified deployment target only applies to the currently compiled module, but not to any client modules, which use inlinable functions from the current module.
Therefore, optimizations which depend on the deployment target must not optimize inlinable functions, except they run after the module has been serialized.

* constant folding of availability checks
* simplification of ARC operations on immortal objects (which depends on the 5.1 runtime)

rdar://88117184

The same thing needs to be done in Sema. There is already a PR for this: https://github.com/apple/swift/pull/33855, but unfortunately currently reverted.

eeckstein committed:
Swift optimizations: make isSwift51RuntimeAvailable sensitive to the resilience domain of the function.

eeckstein committed:
ConstantFolding: don't constant fold availability checks in inlinable functions.

eeckstein commented:
@swift-ci test

eeckstein commented:
@swift-ci test

aschwaighofer commented:
We also need to fix availability guarded prespecialization uses. Something like:
```
--- a/lib/SILOptimizer/Utils/Generics.cpp
+++ b/lib/SILOptimizer/Utils/Generics.cpp
@@ -2512,7 +2512,21 @@ usePrespecialized(SILOptFunctionBuilder &funcBuilder, ApplySite apply,
     auto specializationAvail = SA->getAvailability();
     auto &ctxt = funcBuilder.getModule().getSwiftModule()->getASTContext();
     auto deploymentAvail = AvailabilityContext::forDeploymentTarget(ctxt);
-    auto currentFnAvailability = apply.getFunction()->getAvailabilityForLinkage();
+    auto currentFn = apply.getFunction();
+    auto isInlinableCtxt = currentFn->getResilienceExpansion()
+      == ResilienceExpansion::Minimal;
+    auto currentFnAvailability = currentFn->getAvailabilityForLinkage();
+
+    // If we are in an inlineable function we can't use the specialization except
+    // it the inlinable function itself has availability we can used.
+    if (currentFnAvailability.isAlwaysAvailable() &&
+        isInlinableCtxt) {
+      continue;
+    }
+    else if (isInlinableCtxt) {
+      deploymentAvail = currentFnAvailability;
+    }
+
     if (!currentFnAvailability.isAlwaysAvailable() &&
         !deploymentAvail.isContainedIn(currentFnAvailability))
       deploymentAvail = currentFnAvailability;
```

aschwaighofer commented on lib/SILOptimizer/Utils/ConstantFolding.cpp:
```diff
@@ -1431,6 +1431,9 @@ static bool isApplyOfKnownAvailability(SILInstruction &I) {
   if (!arg2)
     return false;
 
+  if (I.getFunction()->getResilienceExpansion() != ResilienceExpansion::Maximal)
```
Knitpick. Should we move this earlier in this function before the other checks.

aschwaighofer reviewed (approved):
lgtm.

eeckstein commented:
@swift-ci smoke test

eeckstein committed:
Don't use pre-specializations in inlinable functions.

because they need a minimum deployment target, which might not be available in client modules, which de-serialize such functions.

eeckstein commented:
@swift-ci smoke test
